### PR TITLE
BUG: test, fix issue 583

### PR DESCRIPTION
**Pull request by mattip**
fixes issue #583 where ``records([], None, None)`` would raise a ``ValueError``, now it returns an empty record array. Test added that failed, now passes

### CODE REVIEW

File: numpy/core/records.py [hunk 1]
```diff
@@ -661,6 +661,8 @@ def fromrecords(recList, dtype=None, shape=None, formats=None, names=None,
     """
 
     if formats is None and dtype is None:  # slower
+        if len(recList) == 0:
+            return record([])
```
Comment (eric-wieser) [inline numpy/core/records.py]: `record([])` seems broken to me - I'd expect it to return an object with `dtype.names == ()`, but it returns a plain `void`

Comment (mattip) [reply numpy/core/records.py]: good catch, tested and fixed in 44106be

File: numpy/core/records.py [hunk 2]
```diff
@@ -661,6 +661,9 @@ def fromrecords(recList, dtype=None, shape=None, formats=None, names=None,
     """
 
     if formats is None and dtype is None:  # slower
+        if len(recList) == 0:
+            descr = sb.dtype((record, {'names':[], 'formats':[]}))
+            return sb.array([], dtype=descr)
```
Comment (eric-wieser) [inline numpy/core/records.py]: `dtype=(np.record, [])` is enough here.

File: numpy/core/records.py [hunk 3]
```diff
@@ -661,6 +661,9 @@ def fromrecords(recList, dtype=None, shape=None, formats=None, names=None,
     """
 
     if formats is None and dtype is None:  # slower
+        if len(recList) == 0:
```
Comment (eric-wieser) [inline numpy/core/records.py]: I worry this does the wrong thing if `recList.shape = (0, ...)`

File: numpy/core/records.py (hunk 2 above)
Comment (mattip) [reply numpy/core/records.py]: fixed in 564db6f

File: numpy/core/records.py (hunk 3 above)
Comment (mattip) [reply numpy/core/records.py]: Changed in 564db6f . I am not sure what is supposed to happen for different pathological cases. ``np.rec.fromrecords(np.empty((3, 0, 0)))`` still raises an error,  ``np.rec.fromrecords(np.empty((0, 3)))`` does not

File: numpy/core/records.py [hunk 4]
```diff
@@ -662,6 +662,8 @@ def fromrecords(recList, dtype=None, shape=None, formats=None, names=None,
 
     if formats is None and dtype is None:  # slower
         obj = sb.array(recList, dtype=object)
+        if sum(obj.shape) == 0:
```
Comment (eric-wieser) [inline numpy/core/records.py]: Can you maybe just go with the safe `obj.shape == (0,)`, and leave the other cases to be dealt with in future?

Comment (mattip) [reply numpy/core/records.py]: done in 430f598

**Comment by eric-wieser** [comment]
Sorry for flip flopping, but I think that the only acceptable thing to do is error here. Consider:

````
>>> N = 2
>>> np.rec.fromrecords([[1]]*N)
rec.array([(1,), (1,)], 
          dtype=[('f0', '<i4')])
>>> np.rec.fromrecords([1]*N)
rec.array((1, 1), 
          dtype=[('f0', '<i4'), ('f1', '<i4')])
```
When `N == 0`, Is `np.rec.fromrecords([])` a generalization of the first case or the second case? Ie, which of the following is correct?
````
rec.array([], 
          dtype=[])
rec.array((), 
          dtype=[])
```

**Comment by mattip** [comment]
Please document your analysis in the original issue so we can close that as well

**Pull request closed without merging**